File: coffeeshop/src/main/java/com/coffeeshop/coffeeshop/auth/KeycloakTokenClient.java
package com.coffeeshop.coffeeshop.auth;

import org.springframework.http.MediaType;
import org.springframework.stereotype.Component;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.HttpStatusCodeException;
import org.springframework.web.client.RestClient;

import java.util.Map;

@Component
public class KeycloakTokenClient {

    private final KeycloakProperties properties;
    private final RestClient restClient = RestClient.create();

    public KeycloakTokenClient(final KeycloakProperties properties) {
        this.properties = properties;
    }

    public TokenResponse passwordGrant(final String email, final String password) {
        final MultiValueMap<String, String> form = new LinkedMultiValueMap<>();
        form.add("grant_type", "password");
        form.add("client_id", properties.getClientId());
        form.add("client_secret", properties.getClientSecret());
        form.add("username", email);
        form.add("password", password);
        return postToken(form);
    }

    public TokenResponse refreshGrant(final String refreshToken) {
        final MultiValueMap<String, String> form = new LinkedMultiValueMap<>();
        form.add("grant_type", "refresh_token");
        form.add("client_id", properties.getClientId());
        form.add("client_secret", properties.getClientSecret());
        form.add("refresh_token", refreshToken);
        return postToken(form);
    }

    public void logout(final String refreshToken) {
        final MultiValueMap<String, String> form = new LinkedMultiValueMap<>();
        form.add("client_id", properties.getClientId());
        form.add("client_secret", properties.getClientSecret());
        form.add("refresh_token", refreshToken);
        try {
            restClient.post()
                    .uri(properties.logoutUri())
                    .contentType(MediaType.APPLICATION_FORM_URLENCODED)
                    .body(form)
                    .retrieve()
                    .toBodilessEntity();
        } catch (final HttpStatusCodeException ex) {
            throw new KeycloakAuthException("Keycloak logout failed: " + ex.getStatusCode(), ex);
        }
    }

    private TokenResponse postToken(final MultiValueMap<String, String> form) {
        try {
            final Map<?, ?> body = restClient.post()
                    .uri(properties.tokenUri())
                    .contentType(MediaType.APPLICATION_FORM_URLENCODED)
                    .body(form)
                    .retrieve()
                    .body(Map.class);
            if (body == null || body.get("access_token") == null) {
                throw new KeycloakAuthException("Keycloak token response missing access_token");
            }
            final String tokenType = body.get("token_type") != null ? body.get("token_type").toString() : "Bearer";
            final Number expires = (Number) body.get("expires_in");
            final long expiresIn = expires != null ? expires.longValue() : 0L;
            return new TokenResponse(
                    body.get("access_token").toString(),
                    body.get("refresh_token") != null ? body.get("refresh_token").toString() : "",
                    expiresIn,
                    tokenType
            );
        } catch (final HttpStatusCodeException ex) {
            throw new KeycloakAuthException("Keycloak token request failed: " + ex.getStatusCode(), ex);
        }
    }
}

File: coffeeshop-go/internal/auth/keycloak_token_client.go
package auth

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/mastilovic/coffeeshop-go/internal/config"
)

type TokenResponse struct {
	AccessToken  string `json:"access_token"`
	RefreshToken string `json:"refresh_token"`
	ExpiresIn    int64  `json:"expires_in"`
	TokenType    string `json:"token_type"`
}

type KeycloakTokenClient struct {
	cfg        config.Config
	httpClient *http.Client
}

func NewKeycloakTokenClient(cfg config.Config) *KeycloakTokenClient {
	return &KeycloakTokenClient{
		cfg:        cfg,
		httpClient: &http.Client{Timeout: 10 * time.Second},
	}
}

func (c *KeycloakTokenClient) PasswordGrant(email, password string) (*TokenResponse, error) {
	form := url.Values{
		"grant_type":    {"password"},
		"client_id":     {c.cfg.KeycloakClientID},
		"client_secret": {c.cfg.KeycloakClientSecret},
		"username":      {email},
		"password":      {password},
	}
	return c.postToken(form)
}

func (c *KeycloakTokenClient) RefreshGrant(refreshToken string) (*TokenResponse, error) {
	form := url.Values{
		"grant_type":    {"refresh_token"},
		"client_id":     {c.cfg.KeycloakClientID},
		"client_secret": {c.cfg.KeycloakClientSecret},
		"refresh_token": {refreshToken},
	}
	return c.postToken(form)
}

func (c *KeycloakTokenClient) Logout(refreshToken string) error {
	form := url.Values{
		"client_id":     {c.cfg.KeycloakClientID},
		"client_secret": {c.cfg.KeycloakClientSecret},
		"refresh_token": {refreshToken},
	}
	logoutURI := c.cfg.KeycloakBaseURL + "/realms/" + c.cfg.KeycloakRealm + "/protocol/openid-connect/logout"

	resp, err := c.httpClient.PostForm(logoutURI, form)
	if err != nil {
		return fmt.Errorf("keycloak logout request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		return fmt.Errorf("keycloak logout failed with status %d", resp.StatusCode)
	}
	return nil
}

func (c *KeycloakTokenClient) postToken(form url.Values) (*TokenResponse, error) {
	tokenURI := c.cfg.KeycloakBaseURL + "/realms/" + c.cfg.KeycloakRealm + "/protocol/openid-connect/token"

	resp, err := c.httpClient.Post(tokenURI, "application/x-www-form-urlencoded", strings.NewReader(form.Encode()))
	if err != nil {
		return nil, fmt.Errorf("keycloak token request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		return nil, &KeycloakAuthError{StatusCode: resp.StatusCode}
	}

	var body map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&body); err != nil {
		return nil, fmt.Errorf("failed to decode keycloak response: %w", err)
	}

	accessToken, _ := body["access_token"].(string)
	if accessToken == "" {
		return nil, fmt.Errorf("keycloak response missing access_token")
	}

	refreshToken, _ := body["refresh_token"].(string)
	tokenType, _ := body["token_type"].(string)
	if tokenType == "" {
		tokenType = "Bearer"
	}
	var expiresIn int64
	if exp, ok := body["expires_in"].(float64); ok {
		expiresIn = int64(exp)
	}

	return &TokenResponse{
		AccessToken:  accessToken,
		RefreshToken: refreshToken,
		ExpiresIn:    expiresIn,
		TokenType:    tokenType,
	}, nil
}

type KeycloakAuthError struct {
	StatusCode int
}

func (e *KeycloakAuthError) Error() string {
	return fmt.Sprintf("keycloak auth failed with status %d", e.StatusCode)
}
